pydantic_ai_slim/pydantic_ai/models/outlines.py

```diff
@@ -0,0 +1,570 @@
+            hf_tokenizer: The Hugging Face PreTrainedTokenizer or any tokenizer that is compatible with
+                the `transformers` API.
+            provider: The provider to use for OutlinesModel. Can be either the string 'outlines' or an
+                instance of `Provider[OutlinesBaseModel]`. If not provided, the other parameters will be used.
+            profile: The model profile to use. Defaults to a profile picked by the provider.
+            settings: Default model settings for this model instance.
+        """
+        outlines_model: OutlinesBaseModel = from_transformers(hf_model, hf_tokenizer)
+        return cls(outlines_model, provider=provider, profile=profile, settings=settings)
+
+    @classmethod
+    def from_llamacpp(
+        cls,
+        llama_model: 'llama_cpp.Llama',
+        *,
+        provider: Literal['outlines'] | Provider[OutlinesBaseModel] = 'outlines',
+        profile: ModelProfileSpec | None = None,
+        settings: ModelSettings | None = None,
+    ):
+        """Create an Outlines model from a LlamaCpp model.
+
+        Args:
+            llama_model: The llama_cpp.Llama model to use.
+            provider: The provider to use for OutlinesModel. Can be either the string 'outlines' or an
+                instance of `Provider[OutlinesBaseModel]`. If not provided, the other parameters will be used.
+            profile: The model profile to use. Defaults to a profile picked by the provider.
+            settings: Default model settings for this model instance.
+        """
+        outlines_model: OutlinesBaseModel = from_llamacpp(llama_model)
+        return cls(outlines_model, provider=provider, profile=profile, settings=settings)
+
+    @classmethod
+    def from_mlxlm(  # pragma: no cover
+        cls,
+        mlx_model: 'nn.Module',
+        mlx_tokenizer: 'transformers.tokenization_utils.PreTrainedTokenizer',
+        *,
+        provider: Literal['outlines'] | Provider[OutlinesBaseModel] = 'outlines',
+        profile: ModelProfileSpec | None = None,
+        settings: ModelSettings | None = None,
+    ):
+        """Create an Outlines model from a MLXLM model.
+
+        Args:
+            mlx_model: The nn.Module model to use.
+            mlx_tokenizer: The PreTrainedTokenizer to use.
+            provider: The provider to use for OutlinesModel. Can be either the string 'outlines' or an
+                instance of `Provider[OutlinesBaseModel]`. If not provided, the other parameters will be used.
+            profile: The model profile to use. Defaults to a profile picked by the provider.
+            settings: Default model settings for this model instance.
+        """
+        outlines_model: OutlinesBaseModel = from_mlxlm(mlx_model, mlx_tokenizer)
+        return cls(outlines_model, provider=provider, profile=profile, settings=settings)
+
+    @classmethod
+    def from_sglang(
+        cls,
+        openai_client: 'openai.OpenAI | openai.AsyncOpenAI',
```

Comment:
Could it make more sense to provide a base URL and API key instead, and have us build the client from that?

Reply:
I see advantages for both options, but I think I prefer the current one to maintain consistency across the loading methods of the `OutlinesModel` class: all of them require the engine/client from the third party library (just like in Outlines). Otherwise I think that could be a bit confusing for users that `from_openai` behaves differently from all others.